rfcs/ocamlib.md

```diff
@@ -0,0 +1,548 @@
+
+The following behaviours are added to `ocamlobjinfo`.
+
+* The regular `ocamlobjinfo` output on `cmxa` and `cma` is extended
+  according to the current format to output the new `lib_requires` field.
+* `ocamlobjinfo` is made to work on `cmxs` by using the support provided 
+  by `Dynlink` (if available) to output the plugin information rather 
+  than relying on the BDF library (see POC [here][cmxs-read-poc]) 
+* A new `-lib-requires` flag is added that extracts the new `lib_requires`
+  from `cma` and `cmxa` and `dyn_requires` from `cmxs` files. According
+  the following line based format: 
+  ```
+  File: PATH
+  LIBREQ0
+  LIBREQ1
+  LIBREQ2
+  ...
+  ```
+  
+Note this makes `ocamlobjinfo` depend on the C `caml_natdynlink_open`
+primitive investigate if the function is available with an error when
+natdynlink is not available or if we need some kind of conditional
+compilation.
+  
+[cmxs-read-poc]: https://gist.github.com/dbuenzli/0b773f4a4dd9f7a35f30cd9b671e48c5#file-readmeta-ml-L23-L31
+
+## `ocamldep`, `ocamldebug`, `ocamlmktop` and `ocamlmklib` support 
+
+The following behaviours are added to `ocamldep` and `ocamldebug`
+
+* `(ocamldep |ocamldebug) -lib MYLIB`. The repeatable and ordered
+  `-lib MYLIB` option resolves `MYLIB` in `OCAMLPATH` and adds its
+  library directory to includes. Errors if `MYLIB` does not resolve to
+  an existing library.
+
+The following behaviours are added to `ocamlmktop`
+
+* The repeatable and ordered `-lib MYLIB` option resolves and links
+  `MYLIB`'s dependencies as per `ocamlc` support. So goes for direct
+  `cma` arguments.
+
+The following behaviours are added to `ocamlmklib`
+
+* The repeateable and ordered `-lib-require LIB` argument is added 
+  and propogated to the resulting `cma`, `cmxa` and `cmxs` files.
+
+## `ocaml` and `ocamlnat` support
+
+The following behaviours are added to `ocaml` and `ocamlnat`.
+
+* The repeatable and ordered `-lib MYLIB` option resolve 
+  `MYLIB` in `OCAMLPATH` adds its directory to includes and 
+  loads the library and its dependencies according to `cma` linking
+  (for `ocaml`, see `ocamlc` support) or to `cmxs` linking 
+  (for `ocamlnat`, see `ocamlopt` support, though that happens with 
+   the info in `cmxs` files). Note that directories of dependent
+   libraries are *not* added to the includes.
+
+* The new `#require "MYLIB"` directive is added to the interactive toplevel.
+  This directive resolves `MYLIB` in `OCAMLPATH` and adds its library
+  directory to includes. It then loads the library and its
+  dependencies as per the preceeding point. Here again directories of
+  dependent libraries are *not* added to includes.
+ 
+* The `#load` directive is extended to load libraries along the lines of 
+  `#require`.
+ 
+A nice side effect of the new `#require` and `#load` directives is that 
+it doesn't mention file extensions, this allows to use it in `.ocamlinit` 
+and have it work both for `ocaml` and `ocamlnat`.
+
+## `Dynlink` library support
```

Comment:
What about the `findlib.dynload`  mechanism that is used to record the information about the compilation units that are already loaded into the host binary as well as update this information during runtime, when new units are loaded? This mechanism is also reimplemented in Dune. Will this mechanism be totally superseded by this implementation and no longer necessary? (this is what I can conclude from this phrase, but really want to double check):
```
First we check if those exist in the executable itself (cf. linking with `-linkall` above),
if that happens nothing needs to be done.
```

Reply:
> Will this mechanism be totally superseded by this implementation and no longer necessary?

Yes.